manta-trusted-setup/src/ratio.rs

```diff
@@ -0,0 +1,143 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Ratio Proofs
+
+use crate::pairing::{Pairing, PairingEngineExt};
+use ark_ec::{AffineCurve, ProjectiveCurve};
+use ark_ff::{PrimeField, UniformRand, Zero};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, Read, SerializationError, Write};
+use manta_crypto::rand::{CryptoRng, RngCore};
+
+/// Hash to Group Trait for Ratio Proof
+pub trait HashToGroup<P, C>
+where
+    P: Pairing + ?Sized,
+{
+    /// Hashes `challenge` and `ratio` into a group point.
+    fn hash(&self, challenge: &C, ratio: (&P::G1, &P::G1)) -> P::G2;
+}
+
+/// Prepared Ratio Proof Type
+pub type PreparedRatioProof<P> = (
+    (<P as Pairing>::G1Prepared, <P as Pairing>::G1Prepared),
+    (<P as Pairing>::G2Prepared, <P as Pairing>::G2Prepared),
+);
+
+/// Pairing Ratio Proof of Knowledge
+#[derive(derivative::Derivative, CanonicalDeserialize, CanonicalSerialize)]
+#[derivative(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct RatioProof<P>
+where
+    P: Pairing + ?Sized,
+{
+    /// Ratio in G1
+    pub ratio: (P::G1, P::G1),
+
+    /// Matching Point in G2
+    pub matching_point: P::G2,
+}
+
+impl<P> RatioProof<P>
+where
+    P: Pairing + ?Sized,
+{
+    /// Builds a [`RatioProof`] for `scalar` against `challenge`.
+    #[inline]
+    pub fn prove<H, C, R>(
+        hasher: &H,
+        challenge: &C,
+        scalar: &P::Scalar,
+        rng: &mut R,
+    ) -> Option<Self>
+    where
+        H: HashToGroup<P, C>,
+        R: CryptoRng + RngCore + ?Sized,
+    {
+        let g1_point = <P::G1 as AffineCurve>::Projective::rand(rng);
+        if g1_point.is_zero() {
+            return None;
+        }
+        let scaled_g1_point = g1_point.mul(scalar.into_repr());
+        if scaled_g1_point.is_zero() {
+            return None;
+        }
+        let g1_point = g1_point.into_affine();
+        let scaled_g1_point = scaled_g1_point.into_affine();
+        let g2_point = Self::challenge_point(hasher, challenge, (&g1_point, &scaled_g1_point));
+        if g2_point.is_zero() {
+            return None;
+        }
+        let scaled_g2_point = g2_point.mul(*scalar);
+        if scaled_g2_point.is_zero() {
+            return None;
+        }
+        Some(Self {
+            ratio: (g1_point, scaled_g1_point),
+            matching_point: scaled_g2_point.into_affine(),
+        })
+    }
+
+    /// Computes the challenge point that corresponds with the given `challenge`.
+    #[inline]
+    pub fn challenge_point<H, C>(hasher: &H, challenge: &C, ratio: (&P::G1, &P::G1)) -> P::G2
+    where
+        H: HashToGroup<P, C>,
+    {
+        hasher.hash(challenge, (ratio.0, ratio.1))
+    }
+
+    /// Verifies that `self` is a valid ratio proof-of-knowledge, returning the ratio of the
+    /// underlying scalar.
```

Comment:
This is not the ratio proof since it additionally includes `challenge_point`. I have renamed the returned data type from `PreparedRatioProof` to `PreparedRatio`.